### PR TITLE
Don't fail requests if we cannot track their latencies

### LINKED ISSUE #5998: `instrument_dropshot_handler` shouldn't fail if its update fails
Currently:

```rust
        self.update(&context.request, status_code, latency).map_err(|e| {
            HttpError::for_internal_error(format!(
                "error instrumenting dropshot request handler: {}",
                e
            ))
        })?;
        result
```

So even if the actual request succeeds, we may fail if `self.update` fails. That seems like probably the wrong disposition. Practically speaking, I don't see how this could fail, but if it were to fail... what would users do? All the actions they took would be working, but they might not be able to see the effects of those actions. I think ancillary actions such as recording metrics data should steer errors towards an eventual fault management infrastructure and allow client operations to succeed in the face of this internal error.

I got here because I noticed the contents of `nexus/src/external_api/device_auth.rs` has construction an alternate mechanism which, I think, could be avoided if we were to change `instrument_dropshot_handler` to accept a type parameter `E` for the error type. The OAuth entrypoints must return errors in a specific format so may not use `HttpError`.

### PR DESCRIPTION
**Pull request by bnaecker**
Fixes #5998

### CODE REVIEW

File: oximeter/instruments/src/http.rs [hunk 1]
```diff
@@ -93,6 +95,56 @@ impl RequestLatencyHistogram {
     }
 }
 
+/// A helper trait to extract a status code from a Dropshot response.
+pub trait TrackableResponse: HttpResponse {
+    fn status_code(&self) -> StatusCode;
```
Comment (ahl) [inline oximeter/instruments/src/http.rs:100]: What if we add this to dropshot::HttpResponse?

**Review by ahl — APPROVED** [review]

File: oximeter/instruments/src/http.rs (hunk 1 above)
Comment (bnaecker) [reply oximeter/instruments/src/http.rs:100]: That would be nice. The previous code relied on the static method `HttpResponse::response_metadata()`, which doesn't work for a `Response<Body>`, where you need `self` to return the actual status code. Adding that to the existing trait would be neater solution than this one.

Comment (bnaecker) [reply oximeter/instruments/src/http.rs:100]: I've taken this approach in https://github.com/oxidecomputer/dropshot/pull/1049, so I'll pull that update in here, and shrink this PR accordingly.

Comment (bnaecker) [reply oximeter/instruments/src/http.rs]: Updated in 0dee1893d

**Review by ahl — APPROVED** [review]

**Pull request merged**